src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -320,4 +608,102 @@ ErrorMessageOr<RelocatedInstruction> RelocateInstruction(cs_insn* instruction, u
   return result;
 }
 
+uint64_t GetMaxTrampolineSize() {
+  // The maximum size of a trampoline is constant. So the calculation can be cached on first call.
+  static const uint64_t trampoline_size = []() -> uint64_t {
+    MachineCode unused_code;
+    AppendBackupCode(unused_code);
+    AppendPayloadCode(0 /* payload_address*/, 0 /* function address */, unused_code);
+    AppendRestoreCode(unused_code);
+    unused_code.AppendBytes(std::vector<uint8_t>(kMaxRelocatedPrologSize, 0));
+    auto result =
+        AppendJumpBackCode(0 /*address_after_prolog*/, 0 /*trampoline_address*/, unused_code);
+    CHECK(!result.has_error());
+
+    // Round up to the next multiple of 32 so we get aligned jump targets at the beginning of the
+    // each trampoline.
+    return static_cast<uint64_t>(((unused_code.GetResultAsVector().size() + 31) / 32) * 32);
+  }();
+
+  return trampoline_size;
+}
+
+ErrorMessageOr<uint64_t> CreateTrampoline(pid_t pid, uint64_t function_address,
+                                          const std::vector<uint8_t>& function,
+                                          uint64_t trampoline_address, uint64_t payload_address,
+                                          csh capstone_handle,
+                                          absl::flat_hash_map<uint64_t, uint64_t>& relocation_map) {
+  MachineCode trampoline;
+  // Add code to backup register state, execute the payload and restore the register state.
+  AppendBackupCode(trampoline);
+  AppendPayloadCode(payload_address, function_address, trampoline);
+  AppendRestoreCode(trampoline);
+
+  // Relocate prolog into trampoline.
+  OUTCOME_TRY(address_after_prolog,
+              AppendRelocatedPrologCode(function_address, function, trampoline_address,
+                                        capstone_handle, relocation_map, trampoline));
+
+  // Add code for jump from trampoline back into function.
+  OUTCOME_TRY(AppendJumpBackCode(address_after_prolog, trampoline_address, trampoline));
+
+  // Copy trampoline into tracee.
+  auto write_result_or_error =
+      WriteTraceesMemory(pid, trampoline_address, trampoline.GetResultAsVector());
+  if (write_result_or_error.has_error()) {
+    return write_result_or_error.error();
+  }
+
+  return address_after_prolog;
+}
+
+ErrorMessageOr<void> InstrumentFunction(pid_t pid, uint64_t function_address,
```

Comment:
Will the stack be walkable if it we are inside injected code? i.e. does the trampoline have a proper stack frame? Also, I wonder if we can maintain stack-walkability once we also intercept the function exit by overwriting the return address. This is not working on Windows and it's annoying.

Reply:
This refers to stacks produced from sampling, right?

The trampoline is not a function but just a sequence of instructions. It does not get called; we just jump into it. If the unwinding relies on the stack frame the it will look as if the trampoline belongs to the instrumented function. If the unwinder takes the code location into account we will confuse the hell out of it (we live in our own mapping).
Similarly, samples falling into the payload code should be handeled appropriatly (or not, if code location matters).

Overwritten return addresses will pose a problem - I'll think about it again when I have some code for that. I think this is an issue very similar to the problem we have with uretprobes. I talked about this with Davide and there should be solutions for this.

Reply:
Sounds good, thanks.